provisioning/device/samples/solutions/BestPracticeSampleX509/CustomRetryPolicy.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.Azure.Devices.Client.Exceptions;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.Azure.Devices.Client.Samples
+{
+    /// <summary>
+    /// An exponential backoff with jitter based retry policy that retries on additional exceptions
+    /// not covered by the SDK, because in this application we wish to run as long as possible.
+    /// </summary>
+    internal class CustomRetryPolicy : IRetryPolicy
+    {
+        private const int MaxRetryCount = int.MaxValue;
+        private const int MaxExponent = 22; // Avoid integer overlow (max of 30) and clamp max wait to just over 1 hour (2^22 = 1.16 hours).
+
+        private readonly Random _rng = new();
+        private readonly object _randLock = new();
+        private readonly HashSet<Type> _exceptionsToBeRetried;
+
+        private readonly ILogger _logger;
+
+        internal CustomRetryPolicy(HashSet<Type> exceptionsToBeRetried, ILogger logger)
```

Comment:
Why does this sample require a custom retry policy?

Reply:
This was copied from DeviceReconnectionSample. Do you suggest removing this?

Reply:
Per the earlier comment about keeping this sample focused, yes, I would remove most of the superfluous stuff.

Reply:
Wait, can we discuss about this? I think as a best practice, we should include retry and exception handling. The reason is that it is not difficult to implement the happy path. Most of the customer support cases are due to weak handling of exceptions.

Take a look at the customRetry from V1 DeviceReconnectionSample I have patched up for testing with ABB:
https://github.com/Azure/azure-iot-sdk-csharp/tree/andyk/abb

Reply:
> Wait, can we discuss about this?

Of course. That's what we're doing and what I called for in another comment. I'm not trying to dictate here. :)


If every sample includes all best practices, customers can't easily figure out what a specific sample is demonstrating. There is just too much code.

If we did that, we might as well just have 1 uber sample that does everything. What's the point?

The only thing I'd consider adding for this sample (in contrast to what the DeviceReconnectionSample has) is that it is recommended for devices to cache their registration information and try to use it first before calling register again (to avoid dps throttling). That would mean implement part of the connection status callback to know when the client should be reinitialized, and the client re/initialize logic, plus the load-from-cache and save-to-cache logic.